Example server and client: "Forum" / message board

John-Salame opened the pull request:
Branch is based on Yutaka's and currently set to merge into Yutaka's (but it could probably merge directly into main if that is preferable).

I have tested between client and server, using a port that the proxy does not interfere with (port 9090).
How to run: 
python3 example_server_forum.py
python3 example_client_forum.py --ip serverip --port 9090

Current Capabilities: 
1. GET all comments, or comments from a certain user.
2. POST a comment to the server.

Future capabilities (if you think we should add some of them): 
1. Multi-line comments (I think this is currently supported but I never tested it)
3. Select how many comments to show (latest for the selected thread or user)
4. Choose a file to GET from or POST to (instead of having everything in one file).
5. Maybe a limit to the length of data you can POST (for security reasons)

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+def jsonify(user, msg):
+    max_chars = 256
+    num_attributes = 2
+    max_chars -= (num_attributes*6+1) # "":"", and {} but last line has no comma
```
Is max_chars currently used anywhere? If not you can just remove

John-Salame replied:
Good catch, I originally planned to limit the payloads to 256 characters but I don't think I actually used that anywhere.

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+    max_chars = 256
+    num_attributes = 2
+    max_chars -= (num_attributes*6+1) # "":"", and {} but last line has no comma
+    return "{\"user\":\"" + user + "\",\"msg\":\"" + msg + "\"}";
```
This definitely works but can get a little tricky parsing inputs that might also be quoted. Another option is to take a look at the json module: https://docs.python.org/3/library/json.html (e.g. `json.dumps({'user': user, 'msg': msg})`)

John-Salame replied:
Good catch. Maybe I will add something to scrub quotes from key names. If I do add any of the future changes I suggested, I will change this to pass in a dictionary instead of user and message. That sounds like it would actually work perfectly with the json.dumps which you suggested.

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+    print("Comments:")
+    print(r.text)
+    #print("Now printing JSON")
+    #print(r.json())
```
It looks like you can probably drop the commented lines

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+        resp_start = resp[0:json_start]
+        com = r.json()
+        print(com)
+    '''
```
Looks like this string can be removed too

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+    server_address = (ip, ser_port)
+    keep_going = True
+    try:
+        while keep_going:
```
A potential alternative approach is to instead check to see if the user enters get, post, or quit. That would allow the user to gracefully exit the client but remove the continue prompts in between commands.

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+        print('stopping client')
+
+if __name__ == '__main__':
+    if len(sys.argv) != 2:
```
The argument parsing here looks good to me, but just an FYI, there's a python module called argparse that makes it easy to handle more involved cli argument parsing: https://docs.python.org/3/library/argparse.html.

klinvill commented on examples/example_client_forum.py:
```diff
@@ -0,0 +1,77 @@
+    else:
+        ip = sys.argv[1]
+
+    client(ip)
```
I'd recommend making the client also take the port argument. The logic here works, but it's typically good to avoid global variables when possible (unless they're constants)

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+servport = 9090
+posts_file = "./forum/posts.txt"
+
+def jsonify(user, msg):
```
Same comments as for client jsonify. If we had more complex logic, I'd recommend refactoring shared parsing functionality out into a separate module, but it's probably overkill in our case.

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+    return "{\"user\":\"" + user + "\",\"msg\":\"" + msg + "\"}";
+
+# I did this in case any messages have newline characters (so I won't necessarily store comments as one line the posts file)
+def read_simple_json(json_str, start_idx):
```
I'd recommend checking out the json module to see if that makes this easier. Maybe something like: `json.loads(json_str)`

John-Salame replied:
Right now I am still using the read_simple_json, but I can perhaps try switching to json.loads now that I am taking the approach of reading the whole file into a json array (in server.getCommentsFromUser()). What I have right now works for the inputs I've tried, so I'm not too worried about it right now.
Example inputs (for the value of "msg" in the json):
1. {} gonna break? "} \"}
2. What did the cat say to the ghost? "You're creeping meowt!"

My current approach as of 4/23/21 is to use read_simple_json to locate the start and end of a json object in the "posts.txt" file and add a comma to the end of it (before the newline character). Then I sort of turn the key-value pairs into a dictionary and then back into json format. Note that my method would not work for nested JSON objects. I may in the future try to make the read_simple_json just locate the start and end of the object and then call read_simple_json on whatever's in between.
My first priority would be adding an https server to mirror this one, and after I figure that out, then I can go back and see if I can make changes like this.

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+
+        output_txt = ""
+        data_dict = {}
+        try:
```
I don't think you need to put this in a try block since there isn't a following except statement

John-Salame replied:
The "try" block for here is in case there is no question mark (no params in the GET url), index will throw an error. It does run when I send such a request, so I am not sure if I need an except.

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+            # params is everything after "?"
+            params = self.path[self.path.index("?")+1:]
+            data_dict = urllib.parse.parse_qs(params)
+        finally:
```
If you do remove the try statement, then you can also get rid of the finally

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+                print("Number of matches:", cnt)
+                if(cnt == 0):
+                    output_txt = "This user has no comments."
+        except:
```
I'd recommend moving this up to just after the open() call. That way it's clear that the except block will only be triggered if the file open call fails

John-Salame replied:
I believe I also had file reading in the try block in case it fails from a bad / non-open file. I'm looking it up now and it looks like I might be able to use "if not fd.closed" before the rest of the try block to move it out of the "try."
I may try and make this change on my next iteration.

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+            else:
+                comment = ""
+                cnt = 0
+                fd = open(posts_file, "r")
```
The file should already be open so you can omit this line

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+            print("Request params:", data_dict)
+
+        try:
+            fd = open(posts_file, "r")
```
It doesn't really matter in this case, but it's often a good idea to open files in python using the syntax: 
```python
with open(posts_file, "r") as fd:
    fd.read()
```

That syntax makes sure the file is closed once you leave the with statement (without needing to explicitly call close())

John-Salame replied:
Ah, thank you. I learned something new today.
If there is an error opening the file, will it skip over the "with" without problems? I haven't tried it before.

klinvill replied:
Sort of, it will raise an exception like open(). You can wrap it in a try block too though so it could look something like:
```python
try:
    with open(posts_file, "r") as fd:
        contents = fd.read()
except OSError as err : 
    print(err)
    self.send_response(500, message="Server Error")
    return
```

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+                    if not comment:
+                        break
+                    cmt_json = read_simple_json(comment, 0)
+                    #print(cmt_json)
```
Can remove commented lines

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+                cnt = 0
+                fd = open(posts_file, "r")
+                while True:
+                    comment = fd.readline()
```
It looks like comments are delimited by lines. That's totally fine for what we're doing, but if you wanted to store multi-line text, you could instead store and load a list of comments. You could then serialize those comments using the json module. That should generally take care of most parsing issues you'll run into

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+            print(output_txt)
+
+        self.send_response(200, message="Ok")
+        self.send_header("Content-Type", "text/plain")
```
This is totally fine for our purposes, but another alternative could be to send the posts as an application/json content type

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+        self.send_header("Content-Type", "text/plain")
+        self.end_headers()
+        self.wfile.write(output_txt.encode('utf-8'))
+        #self.wfile.write(jsonify("John", "Response").encode('utf-8'))
```
Can remove commented line

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+        print("")
+        print("POST for", self.client_address)
+        print("Request line:", self.requestline)
+        #print(self.headers) # stuff like Host and Content-Type
```
Can remove commented line

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+        #print(self.headers) # stuff like Host and Content-Type
+        data = self.rfile.read(int(self.headers.get('Content-Length'))).decode('utf-8')
+        print(data)
+        #print("Printed POST data")
```
Can remove

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+        #print("Printed POST data")
+
+        # this handles weird url special character stuff, turns params into dict
+        data_dict = urllib.parse.parse_qs(data)
```
I think you can avoid this if we send application/json data. I'm a little surprised that its needed for POST data, but maybe it's because of how text/plain data is encoded.

John-Salame replied:
When I do curl there are no special characters, but the "requests" library on the client side does add special characters into the POST data. It's a bit strange but that's why I ended up adding it.
If you are curious, the content type of the POST requests is "application/x-www-form-urlencoded"

klinvill replied:
Ah ok, that would explain it. The application/x-www-form-urlencoded content type definitely urlencodes the data

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+
+        #write to file
+        fd = open(posts_file, "a")
+        try:
```
I think you can get rid of the try and finally statements here since there's no except clause

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+# https://docs.python.org/3/library/http.server.html
+def httpServer(port):
+    print("Starting HTTP Server")
+    #server_class = socketserver.TCPServer
```
Can remove

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+    print("Starting HTTP Server")
+    #server_class = socketserver.TCPServer
+    server_class = HTTPServer
+    #httpd = socketserver.TCPServer(("", port), MyHandler)
```
Can remove

klinvill commented on examples/example_server_forum.py:
```diff
@@ -0,0 +1,155 @@
+# https://stackoverflow.com/questions/19434947/python-respond-to-http-request
+# https://docs.python.org/3/library/http.server.html
+def httpServer(port):
+    print("Starting HTTP Server")
```
Would be nice to also print the port it's listening on, just as an extra sanity check

klinvill reviewed (changes requested):
Thanks John! It looks pretty good to me but I left a few small comments. Overall though this looks like we should definitely be able to use it to demonstrate our proxies!

John-Salame commented:
Sorry for the inundation of comments, if you got any emails. I don't know how to emphasize a new iteration in the pull request or if this will go to the top of the page, but hopefully it does.
I've addressed most of the concerns from Kirby and added improved file reading logic and client-side visuals. My next goal is to try and make an https server to mirror this one. I haven't done https in Python before (just as I hadn't done http before), so I may need time to figure this out. I'm guessing one goal of the project would be to see if the client proxy can listen on the http destination port and (in Rust that somebody else writes) convert the http messages to an https conversation with the server.

klinvill commented:
> My next goal is to try and make an https server to mirror this one.

I think you can potentially try and follow the approach up here: https://stackoverflow.com/questions/19705785/python-3-simple-https-server. I think this is what you're already planning, but I'd recommend submitting those modifications as a separate pull request and we can merge this PR soon.

klinvill commented on examples/forum/.gitignore:
```diff
@@ -0,0 +1 @@
+__pycache__/*
```
I think we actually want to put this in the root .gitignore. It should also match the __pycache__ directory in forum (in addition to any other __pycache__ directories in our project) if its put there.

klinvill commented on examples/forum/example_client.py:
```diff
@@ -24,9 +19,17 @@ def get(addr):
-    print(r.text)
-    #print("Now printing JSON")
-    #print(r.json())
+    #print(r.text)
```
Can remove this

klinvill commented on examples/forum/example_server.py:
```diff
@@ -0,0 +1,114 @@
+        json_to_send = ut.jsonify_urllib_params(data_dict)
+
+        #write comment in forum post file
+        fd = open(posts_file, "a")
```
I think we also need try except logic here in case we can't open the posts file

klinvill commented on examples/forum/example_server.py:
```diff
@@ -0,0 +1,114 @@
+        except:
+            output_txt = ("\nError opening file \"" + posts_file + "\" or retrieving data.\n")
+            output_txt += "Did you make sure it exists? Maybe post something first.\n"
+            print(output_txt)
```
I didn't mention this in my previous review, but normally if there's an error while processing a request we'll want to send a 500 (Server Error) response. I didn't check, but I think the client might only be processing 200 responses. This is fine for our case, but just more of an FYI

klinvill commented on examples/forum/forum_util.py:
```diff
@@ -0,0 +1,88 @@
+import json
```
I like the util refactoring! 😄

klinvill commented on examples/forum/forum_util.py:
```diff
@@ -0,0 +1,88 @@
+	return json.dumps(new_dict)
+
+# I could use the built-in filter function, but this is a little more clear
+def filter_out_char(in_str, rem_ch):
```
I think you can actually use the replace() function here: `'Test string "with" quotes'.replace('"', '')`

John-Salame replied:
I decided I don't actually need this function for anything so I removed it.

klinvill commented on examples/forum/forum_util.py:
```diff
@@ -0,0 +1,88 @@
+
+
+# I did this in case any messages have newline characters (so I won't necessarily store comments as one line the posts file)
+# Also, the 
```
Incomplete comment, can either remove or add on

klinvill commented on examples/forum/forum_util.py:
```diff
@@ -0,0 +1,88 @@
+		# end of while loop
+	return_dict['_offset'] = offset - start_idx # how many chars to seek to the start of the json
+	return_dict['_length'] = seeker+1 - offset # length of the json in chars
+	return return_dict
```
Can you add a newline at the end of the file? It makes git happy

klinvill reviewed (changes requested):
Thanks for the changes! This looks pretty good. I think there's only one or two comments (in this new review) that need changes, then it looks good to me!

klinvill commented on examples/forum/example_server.py:
```diff
@@ -13,7 +13,13 @@
 def getCommentsFromUser(fd, usr):
     output_txt = "["
-    comment = fd.read()
+    comment = ""
```
This works well but just FYI, IIRC I don't think you actually need this initial assignment here. Try blocks don't have their own inner scope, so any variables you create in there will be accessible outside the block. Since the except block raises an exception, the program will never pass this try block without assigning to comment.

klinvill reviewed (approved):
LGTM! Nice job John and thanks for the changes
